### PR TITLE
My subscriptions error notices

### PR DESCRIPTION
**Pull request by bgrgicak**
### Changes proposed in this Pull Request:

This PR is part of the My subscriptions feature branch.

It introduces standardized success and error notices.
To implement notices we are adding a store and helper functions for adding and removing notices based on the product key.

Error notices are shown on top of the subscriptions list and success notices are shown as snackbars.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Open the [new subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions)
2. Test install, update, and connect
3. Confirm that all return success snackbars
4. To test errors add the code bellow to the `connectProduct`, `installProduct`, and `updateProduct` methods
```
	return new Promise( ( resolve, reject ) =>
		setTimeout(
			() =>
				reject( {
					success: false,

					data: {
						message: 'Failed',
					},
				} ),

			1000
		)
	);
```
5. Test install, update, and connect
6. Confirm that all return error notices

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Add My subscriptions error notices and move existing notices to helper functions.

</details>

### CODE REVIEW

**Review by andfinally — COMMENTED** [review]
I'm getting a 400 error from the install request.

image

I got a "successfully connected" snackbar after updating an active (i.e. connected) product, and afterwards, the product still showed with update available.

image

An update on another product got a 400 error from the request.

image

Connect flow worked as expected.

**Comment by bgrgicak** [comment]
@andfinally  you need to checkout `add/product-slug-to-helper-subscriptions` on WC.test.
Or we need to [merge this PR first](https://github.com/Automattic/woocommerce.com/pull/18534).

File: plugins/woocommerce-admin/client/marketplace/utils/functions.tsx [hunk 1]
```diff
@@ -314,6 +317,31 @@ function installProduct( subscription: Subscription ): Promise< void > {
 	} );
 }
 
+function updateProduct( subscription: Subscription ): Promise< void > {
+	return wpAjax( 'update-' + subscription.product_type, {
+		slug: subscription.local.slug,
+		plugin: subscription.local.path,
+		theme: subscription.local.path,
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/utils/functions.tsx]: It seems odd to pass both these parameters – the product is either a plugin or a theme – but I suppose you have a good reason for it.

File: plugins/woocommerce-admin/client/marketplace/contexts/notice-store.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+							timeout: action.timeout,
+						},
+					},
+				};
+			case 'REMOVE_NOTICE':
+				const notices = { ...state.notices };
+				if ( notices[ action.productKey ] ) {
+					if ( notices[ action.productKey ].timeout ) {
+						clearTimeout( notices[ action.productKey ].timeout );
+					}
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/contexts/notice-store.tsx]: Glad you thought of this!

**Review by andfinally — COMMENTED** [review]
Looking good, thanks. Subscribe/install/update/renew/connect now work for me. I only remembered to check for the snackbars for some of the success messages, but those appeared fine.

I've just realised that you won't see this kind of fixed notice if you have a lot of subs and you're scrolled down the page. Perhaps the best approach to this would be to persist them until dismissed. If that's not easy, we should double or triple the timeout, as they seem to disappear quite quickly.

Is this the right left border colour for an error notice? I thought it would be red, but maybe this is the WP default. Perhaps the WP component has a type prop.

image

I see we're showing `Failed` from the response here. Is there value in showing the API message?

image

File: plugins/woocommerce-admin/client/marketplace/utils/functions.tsx (hunk 1 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/utils/functions.tsx]: Just for simplicity of the implementation. 
I will update it to send just the param it needs.

**Comment by bgrgicak** [comment]
> Perhaps the best approach to this would be to persist them until dismissed. 

We can do that.

>Is this the right left border colour for an error notice? I thought it would be red, but maybe this is the WP default. Perhaps the WP component has a type prop. 

That's odd, it should have been overridden. I will check.

> I see we're showing Failed from the response here. Is there value in showing the API message?

This shouldn't appear.

**Comment by bgrgicak** [comment]
@andfinally I fixed both issues.

> I see we're showing Failed from the response here. Is there value in showing the API message?

This is expected. _Failed_ is the message in the Promise you added in step 4 of the testing instructions.
If you would test with real errors from the API, API error messages would be displayed.

**Review by kdevnel — APPROVED** [review]
Changes look good! The only issue I noticed was the hover state for the "Subscribe button". I couldn't see any others to check if this was universal or not. Either way, seems we're maybe settings a Woo colour somewhere where we should be using the theme colours? 
Screenshot 2023-11-06 at 11 30 29

Also, I did get an error notice at the top and can confirm the colours were correct.

File: plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss [hunk 3]
```diff
@@ -157,3 +160,32 @@
 		margin-right: $grid-unit-10;
 	}
 }
+
+.woocommerce-marketplace__my-subscriptions__notices {
+	.components-notice {
+		margin-left: 0;
+		margin-right: 0;
+		background-color: #fff;
+		box-shadow: 0 2px 6px 0 rgba($gray-100, 0.05);
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss:169]: If this value isn't coming from a design, I would be inclined to darken it a little bit – the notice is hard to spot on the page at the moment.

image

**Review by andfinally — APPROVED** [review]
Thanks for the tweaks, works nicely! 💯 

Tested install, connect - only update as far as the cart page, also tested how two errors behaves.

"Manage on Woo.com" sounds more natural, I'd suggest we use "on" for that option.

image

I experienced the issue you mentioned with host plan subscriptions, where I was able to install a plugin manually and connect a host plan sub for that product. I added a [note](18543-gh-Automattic/woocommerce.com#issuecomment-1795339585) about that to the issue about host plan subs.

> This shouldn't appear.

Sorry! I realised that later, but forgot to update my comment.

File: plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss (hunk 3 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/my-subscriptions/my-subscriptions.scss:169]: It was missing a border. It should look better now.
Screenshot 2023-11-07 at 08 37 13

**Comment by bgrgicak** [comment]
> "Manage on Woo.com" sounds more natural, I'd suggest we use "on" for that option.

Updated

**Pull request merged**